File: packages/typescript/src/distance/distance.ts
import type { Position } from "../types";

const R = 6371.0088;
const DEG = Math.PI / 180;

export function distance(a: Position, b: Position): number {
  const [lng1, lat1] = a;
  const [lng2, lat2] = b;

  const dLat = (lat2 - lat1) * DEG;
  const dLng = (lng2 - lng1) * DEG;

  const sinDLat = Math.sin(dLat / 2);
  const sinDLng = Math.sin(dLng / 2);

  const h =
    sinDLat * sinDLat +
    Math.cos(lat1 * DEG) * Math.cos(lat2 * DEG) * sinDLng * sinDLng;

  return 2 * R * Math.asin(Math.min(1, Math.sqrt(h)));
}

File: packages/java/src/com/peruspatial/distance/Distance.java
package com.peruspatial.distance;

public final class Distance {
  private Distance() {}

  private static final double R = 6371.0088;

  public static double distance(double[] a, double[] b) {
    double lat1 = Math.toRadians(a[1]);
    double lat2 = Math.toRadians(b[1]);
    double dLat = Math.toRadians(b[1] - a[1]);
    double dLng = Math.toRadians(b[0] - a[0]);

    double h = Math.sin(dLat / 2) * Math.sin(dLat / 2)
        + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) * Math.sin(dLng / 2);
    double c = 2 * Math.atan2(Math.sqrt(h), Math.sqrt(1 - h));
    return R * c;
  }
}
